app/scripts/controller-init/snaps/snap-controller-init.ts

```diff
@@ -0,0 +1,121 @@
+import { SnapController } from '@metamask/snaps-controllers';
+import { hasProperty } from '@metamask/utils';
+import { ControllerInitFunction } from '../types';
+import {
+  EndowmentPermissions,
+  ExcludedSnapEndowments,
+  ExcludedSnapPermissions,
+} from '../../../../shared/constants/snaps/permissions';
+import { encryptorFactory } from '../../lib/encryptor-factory';
+import PREINSTALLED_SNAPS from '../../snaps/preinstalled-snaps';
+import { KeyringType } from '../../../../shared/constants/keyring';
+import {
+  SnapControllerInitMessenger,
+  SnapControllerMessenger,
+} from '../messengers/snaps';
+
+/**
+ * Initialize the Snap controller.
+ *
+ * @param request - The request object.
+ * @param request.initMessenger - The init messenger. This has access to
+ * different functions than the controller messenger, and should be used for
+ * initialization purposes only.
+ * @param request.controllerMessenger - The messenger to use for the controller.
+ * @param request.persistedState - The persisted state of the extension.
+ * @param request.removeAllConnections - Function to remove all connections for
+ * a given origin.
+ * @returns The initialized controller.
+ */
+export const SnapControllerInit: ControllerInitFunction<
+  SnapController,
+  SnapControllerMessenger,
+  SnapControllerInitMessenger
+> = ({
+  initMessenger,
+  controllerMessenger,
+  persistedState,
+  removeAllConnections,
+}) => {
+  const allowLocalSnaps = Boolean(process.env.ALLOW_LOCAL_SNAPS);
+  const requireAllowlist = Boolean(process.env.REQUIRE_SNAPS_ALLOWLIST);
+  const rejectInvalidPlatformVersion = Boolean(
+    process.env.REJECT_INVALID_SNAPS_PLATFORM_VERSION,
+  );
+
+  function getMnemonic() {
+    const keyrings = initMessenger.call(
+      'KeyringController:getKeyringsByType',
+      KeyringType.hdKeyTree,
+    );
+
+    if (
+      !keyrings[0] ||
+      !hasProperty(keyrings[0], 'mnemonic') ||
+      !(keyrings[0].mnemonic instanceof Uint8Array)
+    ) {
+      throw new Error('Primary keyring mnemonic unavailable.');
+    }
+
+    // `SnapController` expects a promise.
+    return Promise.resolve(keyrings[0].mnemonic);
+  }
+
+  /**
+   * Get the feature flags for the `SnapController.
+   *
+   * @returns The feature flags.
+   */
+  function getFeatureFlags() {
+    const preferences = initMessenger.call('PreferencesController:getState');
+
+    return {
+      // @ts-expect-error: `useExternalServices` is not defined in the
+      // `PreferencesState` type.
+      disableSnaps: preferences.useExternalServices === false,
```

Comment:
It's in `PreferencesControllerState`? 🤔

Reply:
https://github.com/MetaMask/metamask-extension/blob/fb/snap-ui-renderer-tests/app/scripts/controllers/preferences-controller.ts#L50

Reply:
As far as I can tell we are using the wrong type: https://github.com/MetaMask/metamask-extension/blob/81a49bbb2ce56b74abc861d32dd0280eff1906ac/app/scripts/controller-init/messengers/snaps/snap-controller-messenger.ts#L40 instead of the extension defined one